File: Cpp_logger/metrics/metrics.cpp
#include "metrics.h"

namespace MetricsManager {
    void MetricsManager::HandleGet(const http_request &request) {
        std::cout << "ABOBA" << std::endl;
        request.reply(status_codes::OK, SerializeMetrics());
    }

    MetricsManager::~MetricsManager() {
        Listener.close().wait();
    }

    MetricsManager::MetricsManager(const std::string &path) {
        std::ifstream file(path);

        try {
            Config = web::json::value::parse(file);
        } catch (const web::json::json_exception &e) {
            std::cerr << "Metrics' config parsing error: " << e.what() << std::endl;
            throw;
        }

        file.close();

        Listener = http_listener(
                "http://localhost:" + std::to_string(Config["port"].as_integer()) + Config["metrics_path"].as_string());
        std::cout << "http://localhost:" + std::to_string(Config["port"].as_integer()) + Config["metrics_path"].as_string() << std::endl;
        Listener.support(methods::GET, [this](const http_request &http_req) { this->HandleGet(http_req); });
        Listener.open().wait();
    }

    std::string MetricsManager::SerializeMetrics() {
        std::unique_lock<std::mutex> lock(Mutex);
        if (Config["output_format"].as_string() == "prometheus") {
            std::string result;
            for (const auto& it : Counters) {
                result += it.second->SerializeInPrometheus();
            }
            for (const auto& it : Gauges) {
                result += it.second->SerializeInPrometheus();
            }
            for (const auto& it : Summaries) {
                result += it.second->SerializeInPrometheus();
            }
            return result;
        } else if (Config["output_format"].as_string() == "json") {
            web::json::value result;
            for (const auto& it : Counters) {
                for (const auto& field : it.second->SerializeInJson().as_object()) {
                    result[field.first] = field.second;
                }
            }
            for (const auto& it : Gauges) {
                for (const auto& field : it.second->SerializeInJson().as_object()) {
                    result[field.first] = field.second;
                }
            }
            for (const auto& it : Summaries) {
                for (const auto& field : it.second->SerializeInJson().as_object()) {
                    result[field.first] = field.second;
                }
            }
            return result.serialize();
        }
        std::cerr << "Bad output_format type" << std::endl;
        return "";
    }

    std::shared_ptr<Counter> MetricsManager::GetCounter(const std::string &name) {
        return Counters[name];
    }

    std::shared_ptr<Gauge> MetricsManager::GetGauge(const std::string &name) {
        return Gauges[name];
    }

    std::shared_ptr<Summary> MetricsManager::GetSummary(const std::string &name) {
        return Summaries[name];
    }

    void MetricsManager::RegisterCounter(const std::string& label) {
        if (CheckLabel(label)) {
            Counters[label] = std::make_shared<Counter>(label);
        }
    }

    void MetricsManager::RegisterGauge(const std::string& label) {
        if (CheckLabel(label)) {
            Gauges[label] = std::make_shared<Gauge>(label);
        }
    }

    void MetricsManager::RegisterSummary(const std::string& label, const std::vector<double>& percentiles, size_t cacheSize) {
        if (CheckLabel(label)) {
            Summaries[label] = std::make_shared<Summary>(label, percentiles, cacheSize);
        }
    }

    bool MetricsManager::CheckLabel(const std::string& label) {
        if (Counters.count(label) != 0 && Gauges.count(label) != 0 && Summaries.count(label) != 0) {
            std::cerr << "There is already metric with name " << label << std::endl;
            return false;
        }
        if (label.empty()) {
            std::cerr << "Empty label" << std::endl;
            return false;
        }
        if (!(label.front() >= 'A' && label.front() <= 'Z' ||
              label.front() >= 'a' && label.front() <= 'z')) {
            std::cerr << "Label must begin with a latin letter" << std::endl;
            return false;
        }
        for (const auto& c : label) {
            if (!(c >= 'A' && c <= 'Z' ||
                  c >= 'a' && c <= 'z' ||
                  c >= '0' && c <= '9' ||
                  c == '_')) {
                std::cerr << "Incorrect char in label: " << c << std::endl;
                return false;
            }
        }
        return true;
    }

}

File: java_logger/lib/src/main/java/hse/serrriy/metrics/Metrics.java
package hse.serrriy.metrics;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.function.BiConsumer;

public class Metrics {
    private final HashMap<String, Counter> counters = new HashMap<>();
    private final HashMap<String, Gauge> gauges = new HashMap<>();
    private final HashMap<String, Summary> summaries = new HashMap<>();

    public void createCounter(String name) {
        counters.putIfAbsent(name, new Counter());
    }

    public Counter getCounter(String name) {
        Counter counter = counters.get(name);
        if (counter == null) {
            throw new IllegalArgumentException("Create Counter metric first");
        }
        return counter;
    }

    public void createGauge(String name) {
        gauges.putIfAbsent(name, new Gauge());
    }

    public Gauge getGauge(String name) {
        Gauge gauge = gauges.get(name);
        if (gauge == null) {
            throw new IllegalArgumentException("Create Gauge metric first");
        }
        return gauge;
    }

    public void createSummary(String name, int capacity, double... percentiles) {
        ArrayList<Double> percentilesArray = new ArrayList<>();
        for (double percentile : percentiles) {
            percentilesArray.add(percentile);
        }
        summaries.putIfAbsent(name, new Summary(capacity, percentilesArray));
    }

    public Summary getSummary(String name) throws IllegalArgumentException {
        Summary summary = summaries.get(name);
        if (summary == null) {
            throw new IllegalArgumentException("Create Summary metric first");
        }
        return summary;
    }

    public String getAll() {
        StringBuilder result = new StringBuilder();
        counters.forEach((name, counter) -> {
            result.append(counter.toString(name));
        });
        result.append("\n");
        gauges.forEach((name, gauge) -> {
            result.append(gauge.toString(name));
        });
        result.append("\n");
        summaries.forEach((name, summary) -> {
            result.append(summary.getPercentiles().toString(name));
        });
        return result.toString();
    }
}
